### ISSUE TITLE
Agent prints invalid (too small) set of enable_isntrumentations at the end

### ISSUE THREAD

**Issue by JonasKunz** [elastic/apm-agent-java#2970]
Our current algorithm in `InstrumentationStats` for determining the options to put into `enable_instrumentations` for enabling all used instrumentations is flawed.

For example, running the agent on the following program
```
  @CaptureTransaction
  public static void main(String[] args) throws Exception {
      System.out.println(ElasticApm.currentSpan());
  }
```
yields the following output:
```
co.elastic.apm.agent.bci.InstrumentationStatsLifecycleListener - Used instrumentation groups: [annotations-capture-transaction, executor, executor-collection, fork-join, jul-error, slf4j-error, ssl-context, urlconnection]
```

This set is invalid, because `public-api` would be needed in order to enable the instrumentation for `ElasticApm.currentSpan()`.

Fortunately, the problem we are trying to solve here is a known problem in computer science: [the set cover problem](https://en.wikipedia.org/wiki/Set_cover_problem).
In our case the "universe" is formed by all Instrumentations. Every instrumentation-group-name forms a subset which contains all instrumentations which are enabled by adding this name to `enable_instrumentations`.

I suggest we use two phases of the greedy algorithm for solving the set-cover-problem:
 1. Find a set of instrumentation-group-names which enables all used instrumentations but minimized the number of enabled unused instrumentations
 2. Minimize this set of names while still keeping all used instrumentations enabled

**Comment by eyalkoren** [comment]
I honestly don't see how this is the set cover problem - if you have two instrumentations, each with two groups - `1(A, B)` and `2(A, C)` and only instrumentation `2` is matched, then we want to find the minimal set of groups to represent that. The way to do that is **_not_** finding a minimal collection of **_existing sets_** `(A, B)` and `(A, C)`, in this case it needs to be only `C`, so it looks like a different problem to me. Your universe represents instrumentations and your collection of sets is made up from sets of instrumentation-groups.

That said, I do realize that we have a bug to fix here and I wonder if it is not as simple as adding another step to the algorithm - go again over all used instrumentations and verify that they are represented in the instrumentation group list. If not- simply re-add all groups of that instrumentation.
This guarantees we don't miss. In rare cases the set won't be minimal, but we can definitely live with that. I think it would be rare because AFAIK we normally don't use common instrumentation groups as the entire set, but in combination with specific ones. If it's real easy to implement what you propose (which I need to see code to fully understand) - let's go for that. Otherwise, I don't think this is a core capability so algorithm requirements can be:
1. the output must be such that represents all used instrumentations - strict, that's where it fails now
2. the output should be nearly-minimal - not strict

**Comment by JonasKunz** [comment]
> I honestly don't see how this is the set cover problem - if you have two instrumentations, each with two groups - 1(A, B) and 2(A, C) and only instrumentation 2 is matched, then we want to find the minimal set of groups to represent that. The way to do that is not finding a minimal collection of existing sets (A, B) and (A, C), in this case it needs to be only C, so it looks like a different problem to me. Your universe represents instrumentations and your collection of sets is made up from sets of instrumentation-groups.

You basically need to invert the group-name <-> instrumentation relationship, then it becomes exactly the set-cover problem:
Given the following instrumentations (numbers) with their group names (letters):
 * 1(A, B)
 * 2(A, C)
 * 3(D, B)
 Let 1 and 3 be the used instrumentations, 2 was unused.
 The Universe is now the set of all used isntrumentations: `U={1, 3}`. Each group-name forms a Subset in the Sets *S* with the used instrumentations which is enabled by adding this group-name to `enable_instrumentations` `S = { A={1,3}, B={1}, D={3} }`. This gives the trivial solution of the minimal subset of S : `A={1,3}`. This means we only need to enable the group-name `A` in order to make both 1 & 3 become enabled.
 
 Imo the greedy algorithm from wikipedia is pretty simple to implement, so I would still vote going for that one as it is a known algorithm to solve the problem. My two step variation is only required to minimize the number of accidentally enabled, unused instrumentations, which could be optional but would be nice to optimize startup time.

**Comment by eyalkoren** [comment]
Thanks for the explanation, this is really nice thinking!
However, since you have to do this pre-processing inversion (originally you only have instrumentation-->group mapping and not vice versa), and because you have to use only the used instrumentations for that, you may get the wrong answer.
Extending your example, assume that you also have the following instrumentations:
- 4(B, E)
- 5(B, F)
- 6(B, G)
- 7(B, H)

Your pre-processing inversion will overlook those, because your universe is only made of used instrumentation, therefore you will end up with the same collection of sets: `S = { A={1}, B={1,3}, D={3} }` and your algorithm will therefore choose to enable `B`. However, it is clear that you would be better off with choosing `A` and `D` and avoid using `B`, which is exactly what the current algorithm does.

**Comment by eyalkoren** [comment]
In fact, in real life data, there are typically way more unused instrumentations than used instrumentations. Furthermore, it is likely that instrumentation groups that are common to some are common to many. Therefore, completely ignoring unused instrumentations can lead to these kind of skews.

**Comment by JonasKunz** [comment]
> Your pre-processing inversion will overlook those, because your universe is only made of used instrumentation, therefore you will end up with the same collection of sets: S = { A={1}, B={1,3}, D={3} } and your algorithm will therefore choose to enable B. However, it is clear that you would be better off with choosing A and D and avoid using B, which is exactly what the current algorithm does.

Exactly, using the standard set-cover problem minimizes the number of group-names and ignores the number of matched unused instrumentations.

That's why I was suggesting the two-phase algorithm in my first comment:
> 1. Find a set of instrumentation-group-names which enables all used instrumentations but minimized the number of enabled unused instrumentations
> 2. Minimize this set of names while still keeping all used instrumentations enabled

The first part (1.) is what our current algorithm somehow tries, but doesn't really do it systematically: The algorithm just iterates over the instrumentations in whatever order they happen to lie in memory.

I suggest the following greedy algorithm for (1):
```
Set<String> relevantGroupNames = //get all group names from ALL USED instrumentations
Set<String> enabledGroupNames = {}

while(!allUsedInstrumentationActivatedBy(enabledGroupNames)) {
  String nextGroupToEnable = findGroupNameWhichEnablesMinimumUnusedInstrumentations(relevantGroupNames);
  relevantGroupNames.remove(nextGroupToEnable);
  enabledGroupNames.add(nextGroupToEnable)
}

return enabledGroupNames;
```

This algorithm is guaranteed to enable all used instrumentations and greedly tries to minimize the number of matched unused instrumentations.

This algorithm will naturally return an unnecessary large number of instrumentation group names.
Given the code example from the initial issue comment, the algorithm will 
 1. Add `annotations-capture-transaction`, because it doesn't enable any unused instrumentation
 2. Afterwards, add `public-api` because this is the only way to enable the `ElasticApm.currentSpan()` instrumentation

Clearly, `annotations-capture-transaction` is actually unnecessary becuase `@CaptureTransaction` is already covered by `public-api`. That's why I suggested the set-cover algorithm as second phase in my initial issue comment:

In order to keep the returned set of instrumentation group names now small, we can now further minimize `enabledGroupNames` by applying the standard set-cover algorithm to it (use `enabledGroupNames` to form the subsets as explained in my previous comment).

This would give us a nice, near-optimal list of instrumentation group names.
I think in actual code the algorithm will look much more understandable and more efficient than it seems here, it is a bit hard to explain.

**Comment by eyalkoren** [comment]
OK, now I see where you apply the set cover problem. I think this is where it doesn't really matter anymore 🙂 

> 2. Minimize this set of names while still keeping all used instrumentations enabled

This is not a goal at all! We don't care about the number of instrumentation group names. 

> Given the code example from the initial issue comment, the algorithm will
> 1. Add `annotations-capture-transaction`, because it doesn't enable any unused instrumentation
> 2. Afterwards, add `public-api` because this is the only way to enable the `ElasticApm.currentSpan()` instrumentation
>
> Clearly, `annotations-capture-transaction` is actually unnecessary becuase `@CaptureTransaction` is already covered > by `public-api`. That's why I suggested the set-cover algorithm as second phase in my initial issue comment:

The thing is that enabling `annotations-capture-transaction` and `public-api` is just as good as enabling only `public-api` as it means we only apply the instrumentations we want. There is no cost for more instrumentation groups, there is only cost for more instrumentations.

So I believe your `findGroupNameWhichEnablesMinimumUnusedInstrumentations()` is really the required algorithm. 

Indeed it may improve what we have, but since this is really a very side functionality, I think this is too much effort for marginal improvement. Even this issue, as intellectually interesting as it is, costs more than I would think worth for optimizing the list. 
We need to fix the bug for sure.

With that I think we can conclude the discussion. If you are not convinced - please go ahead and implement it as you wish.

**Comment by JonasKunz** [comment]
I agree with you. For now let's just do your propsoed fix of ensuring that all used instrumentation are covered at the end of the existing algorithm. We can revisit this if we see the requirement for it based on usage (e.g. Lambda cold starts being **regularly** unnecessary slow due to too large `enable_instrumentations` sets).

**Comment by eyalkoren** [comment]
I think we may indeed be required to invest in an improvement for the startup time, but hopefully introducing something that is more efficient and less manual. For example, instrumenting only classes listed in a dedicated file, if such exists, by class name matching only. 
Assuming that the instrumented classes are not expected to change with a combination of agent-build + service-name + service-version, we can rely on a "training run" to create such file. Creating such list in a training run is very easy, robust and deterministic and it represents the absolute minimum matching criteria. 
The main challenge is to decide what qualifies to be a full training run. Ideally, we could have a way to know which classes have not been inspected during the training, thus making it fully automatic and optimizing startup time for all users on all environments. Otherwise, we may offer a manual training run.

Any other idea for automatic or semi-automatic improvement of startup time is more than welcome 🙂
